### DISCUSSION TITLE
Built-in clock overlay

### DISCUSSION THREAD

**Discussion by jgodfrey** [helgeerbe/picframe#130] [Ideas]
I've created a built-in (WIP) clock overlay for picframe that I'd like some feedback on. First, this isn't an attempt to rock the boat in any way with respect to @helgeerbe's overlay work outlined[ in this topic](https://github.com/helgeerbe/picframe/discussions/102). While that work shows a lot of promise for more complex overlays, for a simple clock, I feel that it's a pretty heavy solution and one that'd be more difficult for a casual picture frame user to get up and running.

Additionally, a simple, digital clock is a standard feature on many picture frames, and one I had a personal want for. So, with that in mind, I've created something I'm mostly happy with (though, again, it's still WIP).

Currently, it looks like this:

<a href="https://imgur.com/5xDzJO9"><img src="https://i.imgur.com/5xDzJO9.jpg" title="source: imgur.com" /></a>

Configurable options are:

```
  show_clock: True                        # default=False, True shows clock overlay. False does not show clock overlay
  clock_justify: "R"                      # default="R", clock justification L, C, or R
  clock_text_sz: 120                      # default=120, clock character size
  clock_format: "%I:%M"                   # default="%I:%M", strftime format for clock string
```

Also, I've made it possible to toggle the clock's display `on/off` via MQTT. I haven't yet added it to the HTTP interface, but that should be easy to do.

Also, @paddywwoof, I'd maybe like some input from you on the implementation. Currently, I'm displaying the clock string via a `pi3d.FixedString` object. Since it's not possible (I don't think?) to change the string associated with a `FixedString`, I'm just recreating it as necessary. However, I am being clever enough to only recreate it when necessary (so, when the string changes). In the case of the default time format shown above, it's only necessary to create a new `FixedString` instance once per minute. However, adding a seconds component (for example) to the time format would cause the `FixedString` to get recreated once per second - which still seems reasonable...

It's simple, and seems to work as intended, but if you're aware of a reason I shouldn't do it that way, I'd like to hear it. One reason I went that way is that I'm using the `shadow` ability of the `FixedString` to improve visibility of the text.

Anyway, as I said, I'm still working to clean up a few things but the current, perfectly functional implementation can be found in the [clock-overlay](https://github.com/helgeerbe/picframe/tree/clock-overlay) branch.

Thanks for any input, including whether there's any interest in adding this as a standard feature?

**Comment by sapnho** [comment]
That is pretty neat, I like it a lot!

One idea: Can the clock's color change, like the matting, to always provide a good contrast with the underlying image area?

**Comment by paddywwoof** [comment]
This seems to work fine (tried with various formatting options). The advantage of using FixedString is that the code is basically the same as used just above for the text (so it should be easier to maintain!) I've not tried it on a Pi but there is no noticeable hiccup during blending on the laptop even with seconds - and if it's once a minute it's not going to be an issue, your attention will be distracted by the text change anyway.

The mat colour isn't trivially available in the viewer_display.py file and would only be generated if the  matting option had been chosen by the user. The shadow (as used now) is pretty good but if legibility is a real problem then it might be possible to do something by switching https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glBlendFuncSeparate.xhtml though it's not so obvious what function to choose: making the pixels "negative" of the pixels underneath sounds plausible but the negative of mid grey (127, 127, 127) is mid grey.

PS quick try with:
```python
            pi3d.opengles.glBlendFuncSeparate(pi3d.constants.GL_ONE_MINUS_DST_COLOR, 0, 
                                       1, pi3d.constants.GL_ONE_MINUS_SRC_ALPHA)
            self.__clock_overlay.sprite.draw()
            pi3d.opengles.glBlendFuncSeparate(pi3d.constants.GL_SRC_ALPHA, pi3d.constants.GL_ONE_MINUS_SRC_ALPHA, 
                                       1, pi3d.constants.GL_ONE_MINUS_SRC_ALPHA)
```
has shown me that negative of whatever's underneath is definitely *not* a way to increase legibility!
![Screenshot from 2021-06-25 10-27-23](https://user-images.githubusercontent.com/1939398/123402972-35db9a00-d5a0-11eb-8dba-26cc614e4357.png)
(also issue with alpha blending of edges which would have to be sorted)

PPS maybe using glBlendEquation would be possible but I think it's best to wait util there is a definite problem before trying to fix this.

**Reply by jgodfrey** [reply]
> I've not tried it on a Pi but there is no noticeable hiccup during blending on the laptop even with seconds 

Originally, while working on this, the code designed to cache the previous time value was bugged and it was inadvertently recreating the `FixedString` object for _each frame_. Even then, on my Rpi4, it seemed to run as smooth as ever - which is why I assumed that every minute or even every second would probably be OK.

> That is pretty neat, I like it a lot!

Thanks!

> One idea: Can the clock's color change, like the matting, to always provide a good contrast with the underlying image area?

Interesting idea, though I agree with @paddywwoof's comments above. First, I'm not sure visibility is much of a problem now (with the background shadow). Second, I'm not aware of a good method to guarantee better visibility in all cases (short of rendering the time on top of a (partially transparent) background).

**Comment by helgeerbe** [comment]
I like that clock. Haven't checked the code yes. Is it AM/PM or could you switch to 24 hours?
An additional feature would be some notification icons (mqtt). So I could asked Alexa what's up when coming home (Missed phone calls, washing machine is ready, the mower ist stucked again, ....).

OK, just dreaming.

**Reply by jgodfrey** [reply]
> I like that clock. Haven't checked the code yes. Is it AM/PM or could you switch to 24 hours?

Yeah, you can do that. The configuration file contains the `strftime` configuration value for the clock string - currently set to `%I:%M` (12-hr hour/minute), but you can change it to anything you want, including the formats mentioned above.

> An additional feature would be some notification icons (mqtt). So I could asked Alexa what's up when coming home (Missed phone calls, washing machine is ready, the mower ist stucked again, ....).

Sorry, not sure I follow this? Does this reference the clock somehow or just another potential overlay?

**Reply by helgeerbe** [reply]
Depends on. I just thought of a notification Icon behind the clock time. Could be turned on/off via mqtt.

**Reply by jgodfrey** [reply]
I have already added a clock on/off switch to the MQTT interface, including the Home Assistant code (though I can't test the HA piece). Though, you may be referring to something more than that...

**Reply by jgodfrey** [reply]
FYI - I merged this into `dev` today...

**Reply by helgeerbe** [reply]
I will test with home assistant today.

**Reply by helgeerbe** [reply]
Fixed a typo in clock icon.

Just released 2021.07.05.

Upgrade with `python3 -m pip install -upgrade picframe`
